README.md

```diff
@@ -305,6 +306,30 @@ method takes one argument:
 
 2. `chain` must return a value of the same Chain
 
+### ChainRec
+
+A value that implements the ChainRec specification must also implement the Chain specification.
+
+1. `T.chainRec((done, next, value) => g(done(value)), i) `is equivalent to `g(i)`
```

Comment:
Without putting much thought into it, equivalence looks like it fits the
bill. If anyone disagrees then chime in.

On Sun, 4 Sep 2016, 06:48 Irakli Safareli, <notifications@github.com> wrote:

> In README.md
> <https://github.com/fantasyland/fantasy-land/pull/152#discussion_r77447076>
> :
>
> > @@ -305,6 +306,30 @@ method takes one argument:
> >
> >  2. `chain` must return a value of the same Chain
> >
> > +### ChainRec
> > +
> > +A value that implements the ChainRec specification must also implement the Chain specification.
> > +
> > +1. `T.chainRec((done, next, value) => g(done(value)), i) `is equivalent to `g(i)`
>
> @rpominov <https://github.com/rpominov> type of g is a => T a. maybe I
> have not clearly expressed that with words. it looks like there is no word
> on f/g, for example in monad spec . should I remove the line about g? or
> what you think is best wording for that?
>
> What you think on something like this?
>
> g returns value of type T which contains value of it's first argument
> type.
>
> ------------------------------
>
> About the other laws we could state this
>
> T.chainRec((next,done,v) => p(v) ? f(v).map(done) : g(v).map(next), i) is
> equivalent to
>
> (function step(v) {
>   return p(v) ? f(v) : g(v).chain(step)
> }(i))
>
> But chainRec *must use constant stack space*
>
> And types would be:
>
> p :: a -> boolg :: a -> T bf :: a -> T astep :: a -> T b
>
> And we could also include this in laws like this:
>
> const someNameHere = t => eq => p => f => g => x => {
>     const a = t.chainRec((next,done,v) => p(v) ? f(v).map(done) : g(v).map(next), x);
>     const b = (function step(v) { return p(v) ? f(v) : g(v).chain(step) }(x));
>     return eq(a, b);
> };
>
> But this must be tested with small inputs as step can blow the stack :D
>
> If we call first law identity then what the name should be for this one?
> /cc @SimonRichardson <https://github.com/SimonRichardson>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/fantasyland/fantasy-land/pull/152/files/992f729a420f0ce1343ae67c052b050374329d55#r77447076>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACcaGD2dTr27kK97thrpMa2deKQxuDuUks5qmluqgaJpZM4J0S7m>
> .
>